File: py/git_identify/metadata/github.py
"""
GitHub API metadata extraction.

Fetches file metadata from GitHub's REST API v3.
"""

import json
import os
import urllib.request
from typing import Any, Optional
from urllib.error import HTTPError, URLError

from ..errors import (
    AuthenticationError,
    FileNotFoundError,
    GitError,
    RateLimitError,
    RepositoryNotFoundError,
)
from ..utils.path import normalize_file_path
from ..utils.url import build_github_url


def get_github_metadata(
    owner: str,
    repo: str,
    file_path: str,
    branch: str = "main",
    token: Optional[str] = None
) -> dict[str, Any]:
    """
    Fetch file metadata from GitHub API.

    Retrieves metadata for a file including:
    - Latest commit hash affecting the file
    - File content SHA (blob hash)
    - Last modified timestamp
    - Permalink URL

    Args:
        owner: Repository owner
        repo: Repository name
        file_path: File path relative to repository root
        branch: Branch name (default: 'main')
        token: GitHub personal access token (optional, uses GITHUB_TOKEN env var if not provided)

    Returns:
        Dictionary with normalized metadata

    Raises:
        RepositoryNotFoundError: If repository doesn't exist or is inaccessible
        FileNotFoundError: If file doesn't exist in repository
        AuthenticationError: If authentication fails
        RateLimitError: If API rate limit is exceeded
        GitError: For other API errors

    Examples:
        >>> meta = get_github_metadata("user", "repo", "src/file.py", "main")
        >>> meta['commitHash']
        'abc123def456...'
        >>> meta['fileHash']
        'def456abc789...'
    """
    # Normalize file path
    normalized_path = normalize_file_path(file_path)

    # Get token from parameter or environment
    if token is None:
        token = os.environ.get("GITHUB_TOKEN")

    # Build API URLs
    # 1. Contents API - get file SHA and metadata
    contents_url = (
        f"https://api.github.com/repos/{owner}/{repo}/contents/{normalized_path}"
        f"?ref={branch}"
    )

    # 2. Commits API - get latest commit for file
    commits_url = (
        f"https://api.github.com/repos/{owner}/{repo}/commits"
        f"?path={normalized_path}&sha={branch}&per_page=1"
    )

    try:
        # Fetch file contents metadata
        contents_data = _github_api_request(contents_url, token)

        # Validate response type (should be a file, not a directory)
        if contents_data.get("type") != "file":
            raise FileNotFoundError(
                f"Path is not a file: {file_path}",
                file_path=file_path
            )

        # Fetch commits for this file
        commits_data = _github_api_request(commits_url, token)

        if not commits_data or not isinstance(commits_data, list):
            raise GitError(
                f"No commits found for file: {file_path}",
                context={"file_path": file_path, "branch": branch}
            )

        latest_commit = commits_data[0]

        # Build metadata dictionary
        return {
            "source": "github-api",
            "owner": owner,
            "repo": repo,
            "branch": branch,
            "filePath": normalized_path,
            "commitHash": latest_commit["sha"],
            "fileHash": contents_data["sha"],
            "lastModified": latest_commit["commit"]["committer"]["date"],
            "htmlUrl": build_github_url(owner, repo, latest_commit["sha"], normalized_path)
        }

    except HTTPError as e:
        if e.code == 404:
            if "commits" in e.url:
                raise FileNotFoundError(
                    f"File not found: {file_path}",
                    file_path=file_path,
                    cause=e
                ) from e
            else:
                raise RepositoryNotFoundError(
                    f"Repository not found: {owner}/{repo}",
                    path=f"{owner}/{repo}",
                    cause=e
                ) from e
        elif e.code == 401 or e.code == 403:
            # Check if it's a rate limit error
            if "rate limit" in e.reason.lower():
                reset_time = e.headers.get("X-RateLimit-Reset")
                remaining = e.headers.get("X-RateLimit-Remaining")
                raise RateLimitError(
                    "GitHub API rate limit exceeded",
                    reset_time=int(reset_time) if reset_time else None,
                    remaining=int(remaining) if remaining else 0,
                    cause=e
                ) from e
            else:
                raise AuthenticationError(
                    "GitHub API authentication failed. Please provide a valid GITHUB_TOKEN.",
                    cause=e
                ) from e
        else:
            raise GitError(
                f"GitHub API error: {e.reason}",
                code=f"HTTP_{e.code}",
                context={"url": e.url},
                cause=e
            ) from e

    except URLError as e:
        raise GitError(
            f"Network error accessing GitHub API: {e.reason}",
            code="NETWORK_ERROR",
            cause=e
        ) from e


def _github_api_request(url: str, token: Optional[str] = None) -> Any:
    """
    Make a request to GitHub API.

    Args:
        url: API endpoint URL
        token: GitHub personal access token

    Returns:
        Parsed JSON response

    Raises:
        HTTPError: For HTTP errors
        URLError: For network errors
    """
    # Build request with headers
    headers = {
        "Accept": "application/vnd.github.v3+json",
        "User-Agent": "git-identify-python/2.0.0"
    }

    if token:
        headers["Authorization"] = f"token {token}"

    request = urllib.request.Request(url, headers=headers)

    # Make request
    with urllib.request.urlopen(request, timeout=30) as response:
        data = response.read()
        return json.loads(data.decode("utf-8"))


__all__ = [
    "get_github_metadata",
]

File: mjs/src/metadata/github.mjs
import { normalizeFilePath } from '../utils/path.mjs';
import { buildGitHubUrl } from '../utils/url.mjs';
import {
  FileNotFoundError,
  RepositoryNotFoundError,
  RateLimitError,
  AuthenticationError
} from '../errors.mjs';

/**
 * Gets GitHub API token from environment variable
 * @returns {string | null} GitHub token or null
 */
function getGitHubToken() {
  return process.env.GITHUB_TOKEN || process.env.GH_TOKEN || null;
}

/**
 * Makes a request to GitHub API with authentication and error handling
 * @param {string} url - API endpoint URL
 * @returns {Promise<object>} Response data
 * @throws {Error} Various GitHub API errors
 */
async function githubApiRequest(url) {
  const token = getGitHubToken();
  const headers = {
    'Accept': 'application/vnd.github+json',
    'X-GitHub-Api-Version': '2022-11-28',
    'User-Agent': 'git-identify/2.0.0'
  };

  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }

  const response = await fetch(url, { headers });

  // Handle rate limiting
  if (response.status === 403) {
    const rateLimitRemaining = response.headers.get('x-ratelimit-remaining');
    if (rateLimitRemaining === '0') {
      const resetTime = response.headers.get('x-ratelimit-reset');
      const resetAt = resetTime ? new Date(parseInt(resetTime) * 1000).toISOString() : null;

      throw new RateLimitError(
        'GitHub API rate limit exceeded',
        {
          resetAt,
          context: { url }
        }
      );
    }
  }

  // Handle authentication errors
  if (response.status === 401) {
    throw new AuthenticationError(
      'GitHub API authentication failed. Please set GITHUB_TOKEN environment variable.',
      {
        context: { url }
      }
    );
  }

  // Handle not found
  if (response.status === 404) {
    const data = await response.json().catch(() => ({}));
    throw new FileNotFoundError(
      data.message || 'Resource not found',
      {
        context: { url }
      }
    );
  }

  // Handle other errors
  if (!response.ok) {
    const data = await response.json().catch(() => ({}));
    throw new Error(
      `GitHub API request failed: ${response.status} ${response.statusText}${data.message ? ` - ${data.message}` : ''}`,
    );
  }

  return response.json();
}

/**
 * Retrieves file metadata from GitHub API
 * @param {string} owner - Repository owner
 * @param {string} repo - Repository name
 * @param {string} filePath - File path relative to repository root
 * @param {string} [branch='main'] - Branch name
 * @returns {Promise<object>} Normalized metadata object
 */
export async function getGitHubMetadata(owner, repo, filePath, branch = 'main') {
  if (!owner || !repo || !filePath) {
    throw new TypeError('owner, repo, and filePath are required');
  }

  // Normalize file path
  const normalizedPath = normalizeFilePath(filePath);

  try {
    // Get file contents to retrieve file SHA
    // API: GET /repos/:owner/:repo/contents/:path
    const contentsUrl = `https://api.github.com/repos/${owner}/${repo}/contents/${normalizedPath}?ref=${branch}`;
    const contentsData = await githubApiRequest(contentsUrl);

    // Get commit information for this file
    // API: GET /repos/:owner/:repo/commits
    const commitsUrl = `https://api.github.com/repos/${owner}/${repo}/commits?path=${normalizedPath}&sha=${branch}&per_page=1`;
    const commitsData = await githubApiRequest(commitsUrl);

    if (!commitsData || commitsData.length === 0) {
      throw new FileNotFoundError(
        `No commits found for file "${normalizedPath}" on branch "${branch}"`,
        {
          context: { owner, repo, filePath: normalizedPath, branch }
        }
      );
    }

    const latestCommit = commitsData[0];

    // Build GitHub permalink URL
    const htmlUrl = buildGitHubUrl(owner, repo, latestCommit.sha, normalizedPath);

    // Return normalized metadata
    return {
      source: 'github-api',
      owner,
      repo,
      branch,
      filePath: normalizedPath,
      commitHash: latestCommit.sha,
      fileHash: contentsData.sha,
      lastModified: latestCommit.commit.committer.date,
      htmlUrl
    };
  } catch (error) {
    // Re-throw known errors
    if (error instanceof FileNotFoundError ||
        error instanceof RepositoryNotFoundError ||
        error instanceof RateLimitError ||
        error instanceof AuthenticationError) {
      throw error;
    }

    // Wrap unknown errors
    throw new Error(
      `Failed to retrieve GitHub metadata for ${owner}/${repo}/${normalizedPath}: ${error.message}`,
      { cause: error }
    );
  }
}
